governator-core/src/main/java/com/netflix/governator/internal/PreDestroyMonitor.java

```diff
@@ -0,0 +1,237 @@
+    }
+    
+    private Deque<Callable<Void>> cleanupActions = new ConcurrentLinkedDeque<>();
+    
+    private Map<Integer, ScopeCleanupAction> scopedCleanupActions = new LinkedHashMap<>();
+    private Map<Class<? extends Annotation>, Scope> scopeBindings;
+    private ReferenceQueue<ScopeCleanupMarker> markerReferenceQueue = new ReferenceQueue<>();
+    private final ExecutorService reqQueueExecutor = Executors.newSingleThreadExecutor();
+    private final AtomicBoolean running;
+    
+    private final Provider<ScopeCleanupMarker> markerProvider = new Provider<ScopeCleanupMarker>() {
+        @Override
+        public ScopeCleanupMarker get() {
+            return new ScopeCleanupMarker();
+        }
+    };
+    
+    public PreDestroyMonitor(Map<Class<? extends Annotation>, Scope> scopeBindings) {
+        this.scopeBindings = new HashMap<>(scopeBindings);
+        running = new AtomicBoolean(true);
+        reqQueueExecutor.submit(new Runnable() {
+            public void run() {
+                try {
+                    while (running.get()) {
+                            Reference<? extends ScopeCleanupMarker> ref = markerReferenceQueue.remove(1000);
+                            if (ref != null && ref instanceof ScopeCleanupAction) { 
+                                Integer markerKey = ((ScopeCleanupAction)ref).getId();
+                                ScopeCleanupAction cleanupAction;
+                                synchronized(scopedCleanupActions) {
+                                    cleanupAction = scopedCleanupActions.remove(markerKey);
+                                }
+                                if (cleanupAction != null) {
+                                    cleanupAction.call();
+                                }
+                            }
+                    }
+                } 
+                catch (InterruptedException e) {
+                }                
+            }
+        });
+    }
+    
+    public <T> boolean register(T destroyableInstance, Binding<T> binding, Iterable<LifecycleAction> action) {
+        return binding.acceptScopingVisitor(new ManagedInstanceScopingVisitor(destroyableInstance, binding.getSource(), action));
+    }
+    
+    public void addScopeBindings(Map<Class<? extends Annotation>, Scope> bindings) {
+        scopeBindings.putAll(bindings);
+    }
+
+    /*
+     * compatibility-mode - scope is assumed to be eager singleton
+     */
+    public <T> boolean register(T destroyableInstance, Object context, Iterable<LifecycleAction> action) {
+        return new ManagedInstanceScopingVisitor(destroyableInstance, context, action).visitEagerSingleton();
+    }
+    
+    /**
+     * final cleanup of managed instances if any
+     */
+    @Override
+    public synchronized void close() throws Exception {
+        if (running.compareAndSet(true, false)) {
+            reqQueueExecutor.shutdown();
+            reqQueueExecutor.awaitTermination(2, TimeUnit.SECONDS);
```

Comment:
2 seconds seems very arbitrary and likely to fail for any codebase that has a lengthy shutdown.  We should either make this a configuration option, a much larger number (maybe 10 minutes) or just block indefinitely.